File: src/core/skill-tool.ts
import { tool } from "langchain";
import { z } from "zod";
import type { GraphmindStore } from "../db/client.js";
import type { Skill } from "../types/data-model.js";

/**
 * Format a Skill as a SKILL.md-compatible document following the Agent Skills specification.
 * This ensures skills work identically whether loaded from the graph or from a filesystem.
 *
 * @see https://agentskills.io/specification
 */
export function formatSkillAsMarkdown(skill: Skill): string {
  const frontmatter = [
    "---",
    `name: ${skill.name}`,
    `description: ${skill.description}`,
  ];
  if (skill.tools.length > 0) {
    frontmatter.push(`allowed-tools: ${skill.tools.join(", ")}`);
  }
  if (skill.domain) {
    frontmatter.push(`metadata:`);
    frontmatter.push(`  domain: ${skill.domain}`);
    frontmatter.push(`  confidence: "${skill.confidence.toFixed(2)}"`);
    frontmatter.push(`  trace-count: "${skill.traceCount}"`);
  }
  frontmatter.push("---");

  const body = [
    `# ${skill.name}`,
    "",
    "## Overview",
    "",
    skill.description,
    "",
    "## Instructions",
    "",
    skill.prompt,
  ];

  if (skill.concepts.length > 0) {
    body.push("", `## Tags`, "", skill.concepts.map((c) => `- #${c}`).join("\n"));
  }

  return [...frontmatter, "", ...body].join("\n");
}

/**
 * Creates a `load_skill` tool that agents can use for progressive disclosure.
 *
 * Returns skill content in SKILL.md format (Agent Skills specification),
 * making graph-synthesized skills compatible with the DeepAgents ecosystem.
 *
 * Usage with createAgent():
 * ```typescript
 * const skillTool = createSkillTool(contextGraph.store);
 * const agent = createAgent({
 *   model: "claude-sonnet-4-6",
 *   tools: [...yourTools, skillTool],
 *   middleware: contextGraph.middleware,
 * });
 * ```
 */
export function createSkillTool(store: GraphmindStore) {
  return tool(
    async ({ skill_name }) => {
      // If it's a URL, fetch the remote SKILL.md
      if (skill_name.startsWith("http://") || skill_name.startsWith("https://")) {
        try {
          const response = await fetch(skill_name);
          if (!response.ok) {
            return `Failed to fetch skill from URL: ${response.status} ${response.statusText}`;
          }
          return await response.text();
        } catch (err) {
          return `Failed to fetch skill from URL: ${(err as Error).message}`;
        }
      }

      // Otherwise load from graph
      const skill = await store.getSkillByName(skill_name);

      if (!skill) {
        const available = await store.getSkillsByProject();
        if (available.length === 0) {
          return "No skills available yet. Skills are automatically created as the knowledge lifecycle promotes and clusters decision traces.";
        }
        const manifest = available
          .map((s) => `- ${s.name}: ${s.description} (confidence: ${s.confidence.toFixed(2)})`)
          .join("\n");
        return `Skill "${skill_name}" not found. Available skills:\n${manifest}`;
      }

      return formatSkillAsMarkdown(skill);
    },
    {
      name: "load_skill",
      description:
        "Load a specialized skill by name or URL. " +
        "Pass a skill name to load from the context graph, or a URL to fetch a remote SKILL.md file. " +
        "Use this when the available skills listed in the system prompt match the current task.",
      schema: z.object({
        skill_name: z
          .string()
          .describe(
            "Skill name (e.g., 'handle-account-lockout') or URL to a SKILL.md file"
          ),
      }),
    }
  );
}

/**
 * Creates a `list_skills` tool for discovering available skills.
 */
export function createListSkillsTool(store: GraphmindStore) {
  return tool(
    async () => {
      const skills = await store.getSkillsByProject();
      if (skills.length === 0) {
        return "No skills available yet. Skills are automatically created as decision traces are validated and synthesized.";
      }

      const manifest = skills.map((s) => {
        const tags = s.concepts.length > 0
          ? ` [${s.concepts.map((c) => `#${c}`).join(", ")}]`
          : "";
        const domain = s.domain ? ` (${s.domain})` : "";
        return `- **${s.name}**${domain}: ${s.description} — confidence: ${s.confidence.toFixed(2)}, ${s.traceCount} traces${tags}`;
      });

      return `Available skills:\n${manifest.join("\n")}\n\nUse load_skill to access the full skill context.`;
    },
    {
      name: "list_skills",
      description:
        "List all available skills that can be loaded for specialized guidance. " +
        "Skills are curated bundles of validated decision patterns.",
      schema: z.object({}),
    }
  );
}

/**
 * Export all graph-synthesized skills to a directory as SKILL.md files,
 * compatible with DeepAgents filesystem skills.
 *
 * This bridges graph-based skills with the Agent Skills specification,
 * letting users use synthesized skills with any Agent Skills-compatible framework.
 *
 * ```typescript
 * import { writeFile, mkdir } from "fs/promises";
 *
 * const skills = await contextGraph.store.getSkillsByProject();
 * for (const skill of skills) {
 *   const full = await contextGraph.store.getSkillByName(skill.name);
 *   if (!full) continue;
 *   const dir = `./skills/${skill.name}`;
 *   await mkdir(dir, { recursive: true });
 *   await writeFile(`${dir}/SKILL.md`, formatSkillAsMarkdown(full));
 * }
 * ```
 */
// The formatSkillAsMarkdown function above handles the export format.
// Users call it directly as shown in the example.

File: python-sdk/src/graphmind_context_graphs/core/skill_tool.py
"""Skill tools for progressive disclosure."""

from __future__ import annotations

from langchain.tools import tool
from pydantic import BaseModel, Field

from ..types.data_model import Skill
from ..db.client import GraphmindStore


def format_skill_as_markdown(skill: Skill) -> str:
    lines = ["---", f"name: {skill.name}", f"description: {skill.description}"]
    if skill.tools:
        lines.append(f"allowed-tools: {', '.join(skill.tools)}")
    if skill.domain:
        lines.extend(["metadata:", f"  domain: {skill.domain}", f'  confidence: "{skill.confidence:.2f}"', f'  trace-count: "{skill.trace_count}"'])
    lines.append("---")
    lines.extend(["", f"# {skill.name}", "", "## Overview", "", skill.description, "", "## Instructions", "", skill.prompt])
    if skill.concepts:
        lines.extend(["", "## Tags", "", *(f"- #{c}" for c in skill.concepts)])
    return "\n".join(lines)


class LoadSkillInput(BaseModel):
    skill_name: str = Field(description="Skill name or URL to a SKILL.md file")


def create_skill_tool(store: GraphmindStore):
    @tool("load_skill", args_schema=LoadSkillInput)
    def load_skill(skill_name: str) -> str:
        """Load a specialized skill by name. Use when available skills match the current task."""
        if skill_name.startswith(("http://", "https://")):
            try:
                import urllib.request
                with urllib.request.urlopen(skill_name) as resp:
                    return resp.read().decode()
            except Exception as e:
                return f"Failed to fetch skill from URL: {e}"

        skill = store.get_skill_by_name(skill_name)
        if not skill:
            available = store.get_skills_by_project()
            if not available:
                return "No skills available yet."
            manifest = "\n".join(f"- {s.name}: {s.description} (confidence: {s.confidence:.2f})" for s in available)
            return f'Skill "{skill_name}" not found. Available skills:\n{manifest}'
        return format_skill_as_markdown(skill)

    return load_skill


def create_list_skills_tool(store: GraphmindStore):
    @tool("list_skills")
    def list_skills() -> str:
        """List all available skills that can be loaded for specialized guidance."""
        skills = store.get_skills_by_project()
        if not skills:
            return "No skills available yet."
        items = []
        for s in skills:
            tags = f" [{', '.join(f'#{c}' for c in s.concepts)}]" if s.concepts else ""
            domain = f" ({s.domain})" if s.domain else ""
            items.append(f"- **{s.name}**{domain}: {s.description} — confidence: {s.confidence:.2f}, {s.trace_count} traces{tags}")
        return f"Available skills:\n" + "\n".join(items) + "\n\nUse load_skill to access full skill context."

    return list_skills
